docs/programmers_guide/witness_formal_spec.md

```diff
@@ -0,0 +1,222 @@
+# Block Witness Formal Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://tools.ietf.org/html/rfc2119).
+
+## The Stack
+
+Witnesses are executed on a stack machine that builds tries/calculates hashes.
+
+The stack consists of pairs `(node, hash)`.
+
+Each witness is a queue of instructions.
+
+In every execution cycle a single instruction gets dequeued and a matching substitution rule gets applied to the stack.
+
+In the end, when there are no more instructions left, there MUST be only one item left on the stack.  
+
+## Substitution rules
+
+Here is an example substitution rule. The current instruction is named `INSTR1`.
+
+```
+STACK(node1, hash1) STACK(node2, hash2) INSTR1(params) |=> 
+STACK(helper1(node1, node2), helper2(hash1, hash2))
+```
+
+This substitution rule replaces 2 stack items `(node1, hash1)` and `(node2, hash2)`
+with a single stack item `(helper1(node1, node2), helper2(hash1, hash2))`.
+
+Where `helper1` and `helper2` are pure functions defined in pseudocode (see the example below).
+
+```
+helper1 (value1, value2) {
+    return value1 + value2
+}
+
+helper2 (hash1, hash2) {
+    return KECCAK(hash1 + hash2)
+}
+```
+
+The specification for a substitution rule
+
+```
+[GUARD <CONDITION> ...]
+
+[ STACK(<node-var-name>, <hash-var-name>), ... ] <INSTRUCTION>[(<params>)] |=>
+STACK(node-value, hash-value), ...
+```
+
+The substitution rule MAY have one or more GUARD statements.
+The substitution rule MAY have one or more STACK statements before the instruction.
+The substitution rule MUST have exactly one instruction.
+The substitution rule MAY have parameters for the instruction.
+The substitution rule MUST have at least one STACK statement after the arrow.
+
+So, the minimal substitution rule is for the `HASH` instruction that pushes one hash to the stack:
+```
+HASH(hashValue) |=> STACK(hashNode(hashValue), hashValue)
+```
+
+## GUARDs
+
+Each substitution rule can have zero, one or multiple `GUARD` statements.
+Each `GUARD` statement looks like this:
+
+```
+GUARD <CONDITION>
+```
+
+That means that for the substitution rule to be applicable the `<CONDITION>` in the guard statement must be true.
+
+If a substitution rule has multiple guard statements, all of the conditions specified there should be satisfied.
+
+## `TRAP` statements
+
+If no substitution rules are applicable for an instruction, then the execution MUST stop (`TRAP` instruction) and the partial results MUST NOT be used.
```

Comment:
And what happens if two or more substitution rules are applicable for an instruction?

Reply:
This means the semantics is not valid :)

Reply:
However, it is possible that multiple substitution rules are enabled, but for different instructions. In that case, the order of substitution should not matter

Reply:
we already know which op is current by the opcode, and from a single op's substitution rule there is only one valid rule allowed